src/concurrent.rs

```diff
@@ -146,104 +30,135 @@ pub(crate) struct FilesData {
-    pub(crate) fn new<ProcFiles>(num_jobs: usize, proc_files: ProcFiles) -> Self
+    /// * `num_jobs` - Number of jobs utilized to process files concurrently.
+    pub(crate) fn new<ProcFiles>(
+        proc_files: ProcFiles,
+        num_jobs: usize,
+        complexities: Vec<(Complexity, usize)>,
+    ) -> Self
     where
-        ProcFiles: 'static + Fn(PathBuf, &Config) -> Result<()> + Send + Sync,
+        ProcFiles: 'static + Fn(PathBuf, &[(Complexity, usize)]) -> Option<Snippets> + Send + Sync,
     {
-        let num_jobs = std::cmp::max(2, num_jobs) - 1;
         Self {
             proc_files: Box::new(proc_files),
-            proc_dir_paths: Box::new(null_proc_dir_paths),
-            proc_path: Box::new(null_proc_path),
             num_jobs,
+            complexities,
         }
     }
 
-    /// Runs the producer-consumer approach to process the files
-    /// contained in a directory and in its own subdirectories.
-    ///
-    /// * `config` - Information used to process a file.
-    /// * `files_data` - Information about the files to be included or excluded
-    ///    from a search more the number of paths considered in the search.
-    pub(crate) fn run(
-        self,
-        config: Config,
-        files_data: FilesData,
-    ) -> Result<HashMap<String, Vec<PathBuf>>> {
-        let cfg = Arc::new(config);
-
-        let (sender, receiver) = unbounded();
-
-        let producer = {
-            let sender = sender.clone();
-
-            match thread::Builder::new()
-                .name(String::from("Producer"))
-                .spawn(move || {
-                    explore(
-                        files_data,
-                        &cfg,
-                        self.proc_dir_paths,
-                        self.proc_path,
-                        &sender,
-                    )
-                }) {
-                Ok(producer) => producer,
-                Err(e) => return Err(ConcurrentErrors::Thread(e.to_string()).into()),
+    fn send_file(&self, path: PathBuf, sender: &Sender<PathBuf>) -> Result<()> {
+        sender
+            .send(path)
+            .map_err(|e| Concurrent(format!("Sender: {}", e).into()))
+    }
+
+    fn producer(&self, sender: Sender<PathBuf>, files_data: FilesData) -> Result<()> {
+        let FilesData {
+            path,
+            ref include,
+            ref exclude,
+        } = files_data;
+
+        if !path.exists() {
+            return Err(Concurrent(
+                format!("Sender: {:?} does not exist", path).into(),
+            ));
+        }
+        if path.is_dir() {
+            for entry in WalkDir::new(path)
+                .into_iter()
+                .filter_entry(|e| !is_hidden(e))
+            {
+                let entry = match entry {
+                    Ok(entry) => entry,
+                    Err(e) => return Err(Concurrent(format!("Sender: {}", e).into())),
+                };
+                let path = entry.path().to_path_buf();
```

Comment:
Can we send `&Path` instead of `PathBuf` if we do not modify them or is it impossible?

Reply:
I don't think it's possible because it would introduce lifetime problems, since the lifetime of the reference `&Path` returned by `entry.path()` would not be consistent with the lifetime of an eventual `Sender<&Path>`.